addons/web_editor/static/src/js/editor/snippets.options.js

```diff
@@ -3223,6 +3228,143 @@ registry.background = SnippetOptionWidget.extend({
+            this.$target[0].dataset.shapeData = JSON.stringify(this.initialShape);
+        }
+    },
+
+    //--------------------------------------------------------------------------
+    // Options
+    //--------------------------------------------------------------------------
+
+    /**
+     * @override
+     */
+    setShape(previewMode, widgetValue, params) {
+        if (!previewMode) {
+            this.shape = widgetValue;
+            this._getPositionFromShape();
+            this._setBackground();
+        }
+    },
+    /**
+     * @override
+     */
+    setColor(previewMode, widgetValue, params) {
+        this['color' + params.colorNumber] = this._normalizeColor(widgetValue);
+        this._setBackground();
+    },
+
+    //--------------------------------------------------------------------------
+    // Private
+    //--------------------------------------------------------------------------
+
+    /**
+     * @override
+     */
+    _computeWidgetState(methodName, params) {
+        switch (methodName) {
+            case 'setShape':
+                return this.shape;
+            case 'setColor':
+                return this['color' + params.colorNumber];
+        }
+    },
+    /**
+     * Sets the background to the shape's color-configured url.
+     *
+     * @private
+     */
+    _setBackground() {
+        setBgShape(this.$target[0], !this.shape ? {'background-image': 'none'} : {
+                'background-image': `url("${this._getShapeSrc()}")`,
+                'background-repeat': 'no-repeat',
+                'background-position': this.position,
+                'background-size': '100% auto',
+            },
+        );
+    },
+    /**
+     * Sets the widget's position value by finding the select option
+     * corresponding to the selected shape.
+     *
+     * @private
+     */
+    _getPositionFromShape() {
+        let weButton;
+        if (this.shape) {
+            weButton = this.el.querySelector(`[data-set-shape=${this.shape}]`);
+        }
+        this.position = weButton ? weButton.dataset.position : 'center';
+    },
+    /**
+     * Returns the src of the shape corresponding to the current parameters.
+     *
+     * @private
+     */
+    _getShapeSrc() {
+        return `/website/background_shape/${this.shape}?color_1=${encodeURIComponent(this.color1)}&color_2=${encodeURIComponent(this.color2)}`;
+    },
+    /**
+     * Normalize a color into a css value usable by the canvas rendering context.
+     *
+     * @private
+     * @param {string} color the color to normalize into a css value
+     */
+    _normalizeColor(color) {
+        if (!ColorpickerWidget.isCSSColor(color)) {
+            const style = window.getComputedStyle(document.documentElement);
+            color = style.getPropertyValue('--' + color).trim();
+            color = ColorpickerWidget.normalizeCSSColor(color);
+        }
+        return color;
+    },
```

Comment:
We are multiplying the location where we do that. Could you do a dedicated commit to share that code somewhere?